src/ObjectDetector/index.js

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2020-2024 ml5
+//
+// This software is released under the MIT License.
+// https://opensource.org/licenses/MIT
+
+/*
+ * ObjectDetector: Object detection using cocoSsd in the browser
+ */
+
+import * as tf from "@tensorflow/tfjs";
+import * as cocoSsd from "@tensorflow-models/coco-ssd";
+import callCallback from "../utils/callcallback";
+import handleArguments from "../utils/handleArguments";
+import handleOptions from "../utils/handleOptions";
+import { handleModelName } from "../utils/handleOptions";
+import { mediaReady } from "../utils/imageUtilities";
+
+class ObjectDetector {
+  /**
+   * An object for configuring ObjectDetector options.
+   * @typedef {Object} configOptions
+   * @property {string} modelType - Optional. The type of model to use. Default: "lite".
+   */
+
+  /**
+   * Creates ObjectDetector.
+   * @param {configOptions} options - An object containing ObjectDetector configuration options.
+   * @param {function} callback - A callback to be called when the model is ready.
+   * @private
+   */
+  constructor(modelName, options, callback) {
+    this.modelName = handleModelName(modelName, ["cocoSsd"], "cocoSsd", "objectDetector");
+    this.model = null;
+    this.config = options;
+    this.runtimeConfig = {};
+    this.detectMedia = null;
+    this.detectCallback = null;
+
+    // flags for detectStart() and detectStop()
+    this.detecting = false; // True when detection loop is running
+    this.signalStop = false; // Signal to stop the loop
+    this.prevCall = ""; // Track previous call to detectStart() or detectStop()
+
+    this.ready = callCallback(this.loadModel(), callback);
+  }
+
+  /**
+   * Loads the model.
+   * @return {this} the ObjectDetector model.
+   * @private
+   */
+  async loadModel() {
+    await tf.ready();
+    this.model = await cocoSsd.load();
+    console.log('Finished loading cocoSsd');
+    return this;
+  }
+
+  /**
+   * Asynchronously outputs a single object detection result when called.
+   * Supports both callback and promise.
+   * @param {*} [media] - An HMTL or p5.js image, video, or canvas element to run the detection on.
+   * @param {function} [callback] - Optional. A callback to handle the detection result.
+   * @returns {Promise<Array>} The detection result.
+   */
+  async detect(...inputs) {
+    const argumentObject = handleArguments(...inputs);
+    argumentObject.require("image", "An html or p5.js image, video, or canvas element argument is required for detect().");
+    const { image, callback } = argumentObject;
+
+    await mediaReady(image, false);
+
+    const predictions = await this.model.detect(image);
+    console.log('raw result from cocoSsd', predictions);
```

Comment:
Great for debugging, but normally good to comment this out when sharing this with others (as people are often time confused by console messages that they didn't cause themselves, especially when they are frequent).